File: clients/java/src/test/java/com/google/myanmartools/ZawgyiDetectorTest.java
/* Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.myanmartools;

import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.Truth.assertWithMessage;

import com.google.myanmartools.ZawgyiDetector;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class ZawgyiDetectorTest {

  private static ZawgyiDetector detector;

  @BeforeClass
  public static void setup() {
    detector = new ZawgyiDetector();
  }

  @Test
  public void sanity() {
    double actual = detector.getZawgyiProbability("hello world");
    assertThat(actual).isNegativeInfinity();
  }

  @Test
  public void testIgnoreNonMyanmarCodePoints() {
    String allASCII = "blah blah blah blah blah";
    String mixedUnicode = "<span>blah blah ဒဂုန်ဦးစန်းငွေ </span> blah blah blah blah";
    String mixedZawgyi = "blah blah blah blah blah သို႔သြားပါ။ blah blah blah";
    assertWithMessage("All ASCII")
        .that(detector.getZawgyiProbability(allASCII))
        .isEqualTo(Double.NEGATIVE_INFINITY);
    assertWithMessage("Mixed Unicode")
        .that(detector.getZawgyiProbability(mixedUnicode))
        .isLessThan(0.01);
    assertWithMessage("Mixed Zawgyi")
        .that(detector.getZawgyiProbability(mixedZawgyi))
        .isGreaterThan(0.99);
  }

  @Test
  public void strongUnicodeReturnsLowScore() {
    String strongUnicode = "အပြည်ပြည်ဆိုင်ရာ လူ့အခွင့်အရေး ကြေညာစာတမ်း";
    assertWithMessage("Strong Unicode")
        .that(detector.getZawgyiProbability(strongUnicode))
        .isLessThan(0.001);
  }

  @Test
  public void strongZawgyiReturnsHighScore() {
    String strongZawgyi = "အျပည္ျပည္ဆိုင္ရာ လူ႔အခြင့္အေရး ေၾကညာစာတမ္း";
    assertWithMessage("Strong Zawgyi")
        .that(detector.getZawgyiProbability(strongZawgyi))
        .isGreaterThan(0.999);
  }

  @Test
  public void testDifficult() {
    /* These are strings that the detector has gotten wrong. They mostly render in both Zawgyi and
     * Unicode, but the words they spell make sense in one encoding but not in the other. As the
     * detector improves, change this test case with scores to match the new output.
     */
    Object[][] cases = {
      // STRINGS IDENTICAL IN UNICODE/ZAWGYI
      // TODO: These should return scores in the middle range.
      {"အသံကို အစားထိုးလိုပါသလား။", 0.995}, // Truth: Identical in U/Z
      {"နမူနာ", 0.26}, // Truth: Identical in U/Z
      {" ဦး", 0.35}, // Truth: Identical in U/Z

      // UNICODE STRINGS WITH HIGH ZAWGYI SCORES
      // TODO: Fix detection so that these get low Zawgyi scores.
      {"အစားထိုး အထူးအက္ခရာ", 0.995}, // Truth: Unicode (confirmed by yinmay@)
      {"ယခု မိုးရွာနေပါသလား။", 0.995}, // Truth: Unicode (confirmed by yinmay@)
      {"အခြား", 0.74}, // Truth: Unicode (confirmed by yinmay@)

      // DIFFICULT STRINGS THAT DETECT CORRECTLY
      // Changes to the detector should not significantly change these scores.
      {"ကာမစာအုပ္မ်ား(ေစာက္ပတ္စာအုပ္မ်ား)", 1.0}, // Truth: Zawgyi (confirmed by yinmay@)
      {"ညႇပ္စရာမလို", 0.82}, // Truth: Zawgyi (confirmed by yinmay@)
    };
    for (Object[] cas : cases) {
      String input = (String) cas[0];
      double expected = (double) cas[1];
      double actual = detector.getZawgyiProbability(input);
      assertWithMessage(input).that(actual).isWithin(0.005).of(expected);
    }
  }

  @Test
  public void testIgnoreNumerals() {
    // Digits/numerals are ignored and treated like non-Myanmar code points.
    assertThat(detector.getZawgyiProbability("၉၆.၀ kHz")).isNegativeInfinity();
    assertThat(detector.getZawgyiProbability("၂၄၀၉ ဒဂုန်"))
        .isEqualTo(detector.getZawgyiProbability("ဒဂုန်"));
  }

  @Test
  public void testCompatibility() throws IOException {
    InputStream inputStream = ZawgyiDetector.class.getResourceAsStream("/com/google/myanmartools/compatibility.tsv");
    BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, "UTF-8"));
    String line = null;
    while ((line = reader.readLine()) != null) {
      int tabIdx = line.indexOf('\t');
      double expected = Double.parseDouble(line.substring(0, tabIdx));
      String input = line.substring(tabIdx + 1);
      double actual = detector.getZawgyiProbability(input);
      try {
        // Java Truth has very limited support for double equality testing. Prior to JDK 11, this
        // test passed for both infinite and non-infinite with .isEqualTo(), but now it fails with
        // a small delta.
        if (Double.isInfinite(expected)) {
          assertWithMessage(line).that(actual).isEqualTo(expected);
        } else {
          assertWithMessage(line).that(actual).isWithin(1e-6).of(expected);
        }
      } catch (AssertionError e) {
        // Print verbose output for the failure
        detector.getZawgyiProbability(input, true);
        throw e;
      }
    }
    reader.close();
  }
}

File: clients/c#/MyanmarTools.Test/ZawgyiDetectorTest.cs
using Xunit;
using System;
using System.IO;
using System.Text;

namespace MyanmarTools.Test
{

    public class ZawgyiDetectorTest
    {
        private ZawgyiDetector detector;
        public ZawgyiDetectorTest()
        {
            detector = new ZawgyiDetector();
        }

        [Fact(DisplayName = "Sanity")]
        public void Sanity()
        {
            double actual = detector.GetZawgyiProbability("hello world");
            Assert.Equal(actual, double.NegativeInfinity);
        }

        [Fact(DisplayName = "testIgnoreNonMyanmarCodePoints")]
        public void testIgnoreNonMyanmarCodePoints()
        {
            var allASCII = "blah blah blah blah blah";
            var mixedUnicode = "<span>blah blah ဒဂုန်ဦးစန်းငွေ </span> blah blah blah blah";
            string mixedZawgyi = "blah blah blah blah blah သို႔သြားပါ။ blah blah blah";

            //All ASCII
            var Expected = double.NegativeInfinity;
            var Actual = detector.GetZawgyiProbability(allASCII);
            Assert.True(Actual == Expected, $"Error:All ASCII; Expected:{Expected} == Actual:{Actual}");

            //Mixed Unicode
            Expected = 0.01;
            Actual = detector.GetZawgyiProbability(mixedUnicode);
            Assert.True(Actual < Expected, $"Error:Mixed Unicode; Expected:{Expected} < Actual:{Actual}");

            //Mixed Zawgyi
            Expected = 0.99;
            Actual = detector.GetZawgyiProbability(mixedZawgyi);
            Assert.True(Actual > 0.99, $"Error:Mixed Zawgyi; Expected:{Expected} > Actual:{Actual}");

        }
        [Fact(DisplayName = "strongUnicodeReturnsLowScore")]
        public void strongUnicodeReturnsLowScore()
        {
            var strongUnicode = "အပြည်ပြည်ဆိုင်ရာ လူ့အခွင့်အရေး ကြေညာစာတမ်း";

            var Expected = 0.001;
            var Actual = detector.GetZawgyiProbability(strongUnicode);
            Assert.True(Actual < Expected, $"Error:Strong Unicode; Expected:{Expected} < Actual:{Actual}");

        }

        [Fact(DisplayName = "strongZawgyiReturnsHighScore")]
        public void strongZawgyiReturnsHighScore()
        {
            var strongZawgyi = "အျပည္ျပည္ဆိုင္ရာ လူ႔အခြင့္အေရး ေၾကညာစာတမ္း";
            var Expected = 0.999;
            var Actual = detector.GetZawgyiProbability(strongZawgyi);
            Assert.True(Actual > Expected, $"Error:Strong Zawgyi; Expected:{Expected} > Actual:{Actual}");

        }

        [Theory(DisplayName = "testDifficult")]
        // STRINGS IDENTICAL IN UNICODE/ZAWGYI
        // TODO: These should return scores in the middle range.
        [InlineData("အသံကို အစားထိုးလိုပါသလား။", 0.995)] // Truth: Identical in U/Z
        [InlineData("နမူနာ", 0.26)] // Truth: Identical in U/Z
        [InlineData(" ဦး", 0.35)] // Truth: Identical in U/Z
        // UNICODE STRINGS WITH HIGH ZAWGYI SCORES
        // TODO: Fix detection so that these get low Zawgyi scores.
        [InlineData("အစားထိုး အထူးအက္ခရာ", 0.995)] // Truth: Unicode
        [InlineData("ယခု မိုးရွာနေပါသလား။", 0.995)] // Truth: Unicode
        [InlineData("အခြား", 0.74)] // Truth: Unicode
        // DIFFICULT STRINGS THAT DETECT CORRECTLY
        // Changes to the detector should not significantly change these scores.
        [InlineData("ကာမစာအုပ္မ်ား(ေစာက္ပတ္စာအုပ္မ်ား)", 1.0)] // Truth: Zawgyi
        [InlineData("ညႇပ္စရာမလို", 0.82)] // Truth: Zawgyi
        public void testDifficult(string Input, double Expected)
        {
            var Actual = detector.GetZawgyiProbability(Input);
            var ExpectedDiff = 0.005;
            var ActualDiff = Math.Abs(Actual - Expected);
            Assert.True(ActualDiff <= ExpectedDiff, $"Error:{Input}; Expected:{ExpectedDiff} Different Actual:{ActualDiff}");
        }

        [Fact(DisplayName = "testIgnoreNumerals")]
        public void testIgnoreNumerals()
        {
            // Digits/numerals are ignored and treated like non-Myanmar code points.
            Assert.Equal(detector.GetZawgyiProbability("၉၆.၀ kHz"), double.NegativeInfinity);
            Assert.Equal(detector.GetZawgyiProbability("၂၄၀၉ ဒဂုန်"), detector.GetZawgyiProbability("ဒဂုန်"));

        }

        [Fact(DisplayName = "testCompatibility")]
        public void testCompatibility()
        {
            var line = "";
            using (var steam = File.Open(Path.Combine("resources", "compatibility.tsv"), FileMode.Open, FileAccess.Read, FileShare.Read))
            {
                using (var streamReader = new StreamReader(steam, Encoding.UTF8, true, 1024))
                {
                    while (streamReader.Peek() != -1)
                    {
                        line = streamReader.ReadLine();
                        int tabIdx = line.IndexOf('\t');
                        var strValue=line.Substring(0, tabIdx);
                        double Expected = double.NegativeInfinity;
                        if (strValue != "-Infinity")
                        {
                           Expected= double.Parse(strValue);
                        }
                        var input = line.Substring(tabIdx + 1);
                        double Actual = detector.GetZawgyiProbability(input);

                        if (Expected == double.NegativeInfinity) { 
                          Assert.True(Actual==Expected,$"Error:{input}; Expected:{strValue} Different Actual:{Actual}");
                        }else { 
                          var ActualDiff = Math.Abs(Actual - Expected);
                           Assert.True(ActualDiff<=1e-6,$"Error:{input}; Expected:{strValue} Different Actual:{Actual}");
                        }
                    }

                }
            }

        }



    }


}


